clientgui/MainDocument.cpp

```diff
@@ -1750,61 +1795,89 @@ int CMainDocument::WorkShowGraphics(RESULT* rp) {
 #ifdef __WXMSW__
         HANDLE   id;
 #else
-        int      id;
+        int      id = 0;
 #endif
 
         // See if we are already running the graphics application for this task
         previous_gfx_app = GetRunningGraphicsApp(rp);
 
         if (previous_gfx_app) {
-            // If graphics app is already running, the button has changed to
-            // "Stop graphics", so we end the graphics app.
-            //
-            KillGraphicsApp(previous_gfx_app->pid); // User clicked on "Stop graphics" button
+#ifdef __APPLE__
+            if (g_use_sandbox) {
+                // If graphics app is already running, the button has changed to
+                // "Stop graphics", so we end the graphics app.
+                //
+                // For some graphics apps (including Einstein), we must explicitly delete both
+                // our forked process and its child graphics app, or the other will remain. A
+                // race condition can occur because of the time it takes for our forked process
+                // to launch the graphics app, so we periodically poll for the child's PID until
+                // it is available. If we don't have it yet, then defer responding to the "Stop 
+                // graphics" button.
+                if (GetGFXPIDFromForkedPID(previous_gfx_app)) {
+                    KillGraphicsApp(previous_gfx_app->pid); // User clicked on "Stop graphics" button
+                    KillGraphicsApp(previous_gfx_app->gfx_pid);
+                }
+            } else {
+                KillGraphicsApp(previous_gfx_app->pid); // User clicked on "Stop graphics" button
+            }
+#else
+        // On other platforms, don't launch a second instance if graphics app is already running
```

Comment:
Shouldn't we kill already running graphics app on other platforms as well?

Reply:
The code is correct as written.

This part of the code is refactored from lines 1799-1801 of _Maindocument.cpp_ in GIT master. On other platforms, it is executed only when the user clicks on "Show graphics" in BOINC Manager for a selected task; if the graphics for the selected task is already running, we just want to ignore this click. 

On Mac only, this button is changed to "Stop graphics" when the task is running, so we want to kill it on Mac only. I made the button change to "Stop graphics" is that some project graphics apps (e.g., Einstein) don't respond to the close box on newer versions of MacOS.